File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No59SpiralMatrixII.java
package com.hazza.LeetCodeByJava.SegmentTwo;

/**
 * Created with Intellii IDEA.
 * Description: https://leetcode.com/problems/spiral-matrix-ii/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-14
 * Time: 8:38 AM
 */
public class No59SpiralMatrixII {
    public int[][] generateMatrix(int n) {
        if (n <= 0) return new int[0][0];
        int rowBegin = 0, rowEnd = n - 1, colBegin = 0, colEnd = n - 1;
        int count = 1;
        int[][] res = new int[n][n];
        while (count <= n * n) {
            // left to right
            for (int i = colBegin; i <= colEnd; i++, count++)
                res[rowBegin][i] = count;
            ++rowBegin;
            // top to down
            for (int i = rowBegin; i <= rowEnd; i++, count++)
                res[i][colEnd] = count;
            --colEnd;
            // right to left
            if (rowEnd >= rowBegin) {
                for (int i = colEnd; i >= colBegin; i--, count++)
                    res[rowEnd][i] = count;
            }
            --rowEnd;
            // down to top
            if (colEnd >= colBegin) {
                for (int i = rowEnd; i >= rowBegin; i--, count++)
                    res[i][colBegin] = count;
            }
            ++colBegin;
        }
        return res;
    }
}

File: src/main/python/leetcode_by_python/SegmentTwo/No59_Spiral_Matrix_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/9 20:06
# https://leetcode.com/problems/spiral-matrix-ii/
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0] * n for _ in range(n)]
        i, j, di, dj = 0, 0, 0, 1

        for k in range(1, n*n+1):
            res[i][j] = k
            if res[(i+di)%n][(j+dj)%n]:
                di, dj = dj, -di
            i += di
            j += dj

        return res


if __name__ == '__main__':
    solution = Solution()

    expected = [
        [1, 2, 3],
        [8, 9, 4],
        [7, 6, 5]
    ]
    assert solution.generateMatrix(3) == expected
